File: 207-course-schedule/course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:

        
        
        adj = [[] for num in range(numCourses)]

        visited = [0 for num in range(numCourses)]

        curr_path = [0 for num in range(numCourses)]
        
        for r in prerequisites:
            adj[r[1]].append(r[0])

        def dfs(curr):
            if curr_path[curr] == 1:
                return False
            
            if visited[curr] == 1:
                return True
            visited[curr] = 1
            curr_path[curr] = 1

            for n in adj[curr]:
                if not dfs(n):
                    return False

            curr_path[curr] = 0
            return True
        
        for n in range(numCourses):
            if visited[n] == 0:
                if not dfs(n):
                    return False

        
        return True

File: 207-course-schedule/course-schedule.cpp
class Solution {
public:
    bool bfs(vector<vector<int>>& graph, int V,vector<int>& indegree, vector<int>& visited){
        queue<int> q;
        vector<int> ans;
        for(int vertex = 0; vertex < V; vertex++){
            if (indegree[vertex] == 0){
                q.push(vertex);
                ans.push_back(vertex);
                visited[vertex] = 1;
            }
        }
        while(!q.empty()){
            int vertex = q.front();
            q.pop();
            for(int neighbor: graph[vertex]){
                indegree[neighbor] -= 1;
                if (indegree[neighbor] == 0){
                    ans.push_back(neighbor);
                    q.push(neighbor);
                    visited[neighbor] = 1;
                }
            }
        }
        return ans.size() == V;
    }
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<int> indegree(numCourses);
        vector<vector<int>> graph(numCourses);
        vector<int> visited(numCourses);
        for(vector<int> connection: prerequisites){
            int start = connection[0];
            int end = connection[1];
            indegree[end]++;
            graph[start].push_back(end);
        }
        return bfs(graph, numCourses, indegree, visited);
    }
};